onnxruntime/core/providers/cpu/ml/treeregressor.cc

```diff
@@ -239,40 +287,172 @@ common::Status TreeEnsembleRegressor<T>::Compute(OpKernelContext* context) const
   int64_t N = X->Shape().NumDimensions() == 1 ? 1 : X->Shape()[0];
   Tensor* Y = context->Output(0, TensorShape({N, n_targets_}));
 
-  int64_t write_index = 0;
   const auto* x_data = X->template Data<T>();
 
-  for (int64_t i = 0; i < N; i++)  //for each class
-  {
-    int64_t current_weight_0 = i * stride;
-    std::unordered_map<int64_t, std::tuple<float, float, float>> scores; // sum, min, max
-    //for each tree
-    for (size_t j = 0; j < roots_.size(); j++) {
-      //walk each tree from its root
-      ORT_RETURN_IF_ERROR(ProcessTreeNode(scores, roots_[j], x_data, current_weight_0));
+  if (n_targets_ == 1) {
+    float origin = base_values_.size() == 1 ? base_values_[0] : 0.f;
+    if (N == 1) {
+      float scores = 0;
+      unsigned char has_scores = 0;
+
+      if (nbtrees_ >= 2 && (aggregate_function_ == AGGREGATE_FUNCTION::AVERAGE ||
+                            aggregate_function_ == AGGREGATE_FUNCTION::SUM)) {
+#ifdef USE_OPENMP
+#pragma omp parallel for reduction(|                         \
+                                   : has_scores) reduction(+ \
+                                                           : scores)
+#endif
+        for (int64_t j = 0; j < nbtrees_; ++j)
```

Comment:
We're going to try and replace this sort of thing with calls to ThreadPool::[Batch]ParallelFor in the near future so that the threading is more consistent (gets confusing when you can have the intra-op and inter-op threadpools using Eigen threads, and places like this using OpenMP). Not sure how easy it is to convert this sort of pragma, but it would be good to try.

The intra-op threadpool can be obtained from OpKernelContext::GetOperatorThreadPool.